contracts/ronin/slash-indicator/SlashFastFinality.sol

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.9;
+
+import "../../interfaces/slash-indicator/ISlashFastFinality.sol";
+import { IRoninValidatorSet } from "../../interfaces/validator/IRoninValidatorSet.sol";
+import { IProfile } from "../../interfaces/IProfile.sol";
+import { IRoninTrustedOrganization } from "../../interfaces/IRoninTrustedOrganization.sol";
+import "../../precompile-usages/PCUValidateFastFinality.sol";
+import "../../extensions/collections/HasContracts.sol";
+import "../../utils/CommonErrors.sol";
+
+abstract contract SlashFastFinality is ISlashFastFinality, HasContracts, PCUValidateFastFinality {
+  /// @dev The amount of RON to slash fast finality.
+  uint256 internal _slashFastFinalityAmount;
+  /// @dev The block number that the punished validator will be jailed until, due to malicious fast finality.
+  uint256 internal _fastFinalityJailUntilBlock;
+  /// @dev Recording of submitted proof to prevent relay attack.
+  mapping(bytes32 => bool) internal _processedEvidence;
+
+  /**
+   * @dev This empty reserved space is put in place to allow future versions to add new
+   * variables without shifting down storage in the inheritance chain.
+   */
+  uint256[21] private ______gap;
+
+  modifier onlyGoverningValidator() {
+    if (_getGovernorWeight(msg.sender) == 0) revert ErrUnauthorized(msg.sig, RoleAccess.GOVERNOR);
+    _;
+  }
+
+  /**
+   * @inheritdoc ISlashFastFinality
+   */
+  function slashFastFinality(
+    address consensusAddr,
+    bytes memory voterPublicKey,
+    uint256 targetBlockNumber,
+    bytes32[2] memory targetBlockHash,
+    bytes[][2] memory listOfPublicKey,
+    bytes[2] memory aggregatedSignature
```

Comment:
```suggestion
    bytes calldata voterPublicKey,
    uint256 targetBlockNumber,
    bytes32[2] calldata targetBlockHash,
    bytes[][2] calldata listOfPublicKey,
    bytes[2] calldata aggregatedSignature
```

Reply:
Please write an explanation for this suggestion.

Reply:
Since Calldata has already been paid for with the transaction, if you don’t modify a parameter to a function, then don’t bother copying the function to memory and just read the value from calldata.